From f2c05895419aedd00cda8dfe06c1ad50f706ff58 Mon Sep 17 00:00:00 2001
From: Davide Beatrici <git@davidebeatrici.dev>
Date: Tue, 6 Jan 2026 03:03:32 +0100
Subject: [PATCH] feat: allow to use current working directory as the agent's

This is the case in OpenMandriva's package for example:

1. Tanuki Wrapper is built separately in its own dedicated package and doesn't live in the agent's directory.
2. The agent JAR is outside the lib folder.
---
 src/main/java/io/onedev/agent/Agent.java | 49 +++++++++++++++---------
 1 file changed, 31 insertions(+), 18 deletions(-)

diff --git a/src/main/java/io/onedev/agent/Agent.java b/src/main/java/io/onedev/agent/Agent.java
index 2765398..af7f652 100644
--- a/src/main/java/io/onedev/agent/Agent.java
+++ b/src/main/java/io/onedev/agent/Agent.java
@@ -147,23 +147,7 @@ public class Agent {
 		}
 		
 		try {
-			sandboxMode = new File("target/sandbox").exists();
-	
-			String path = Agent.class.getProtectionDomain().getCodeSource().getLocation().toURI().getPath();
-			File loadedFrom = new File(path);
-			
-			if (loadedFrom.getParentFile() != null 
-					&& loadedFrom.getParentFile().getParentFile() != null
-					&& loadedFrom.getParentFile().getParentFile().getParentFile() != null
-					&& new File(loadedFrom.getParentFile().getParentFile().getParentFile(), "conf/agent.properties").exists()) {
-				installDir = loadedFrom.getParentFile().getParentFile().getParentFile();
-			} else if (new File("target/sandbox").exists()) {
-				installDir = new File("target/sandbox");
-			} else {
-				throw new RuntimeException("Unable to find agent directory");
-			}
-	
-			installDir = installDir.getCanonicalFile();
+			installDir = getInstallDir();
 
 			File testFile = new File(installDir, "test");
 			try (var os = new FileOutputStream(testFile)) {
@@ -543,7 +527,36 @@ public class Agent {
 	public static File getWorkDir() {
 		return new File(installDir, "work");
 	}
-	
+
+	public static File getInstallDir() {
+		try {
+			String agentPropsPath = "conf/agent.properties";
+
+			File file = new File(agentPropsPath);
+			if (file.exists()) {
+				return new File(".").getCanonicalFile();
+			}
+
+			String loadPath = Agent.class.getProtectionDomain().getCodeSource().getLocation().toURI().getPath();
+			file = new File(loadPath);
+			if (file.getParentFile() != null
+					&& file.getParentFile().getParentFile() != null
+					&& file.getParentFile().getParentFile().getParentFile() != null
+					&& new File(file.getParentFile().getParentFile().getParentFile(), agentPropsPath).exists()) {
+				return file.getParentFile().getParentFile().getParentFile().getCanonicalFile();
+			}
+
+			file = new File("target/sandbox");
+			if (file.exists()) {
+				return file.getCanonicalFile();
+			}
+		} catch (Exception e) {
+			logger.error("Error getting install directory", e);
+		}
+
+		throw new RuntimeException("Unable to find agent directory");
+	}
+
 	public static void log(Session session, String jobToken, String message, @Nullable String sessionId) {
 		if (sessionId == null)
 			sessionId = "";
-- 
2.51.2

